stdlib/README.adoc

```diff
@@ -0,0 +1,208 @@
+  where we need to keep something in the stdlib for backward
+  compatibility while knowing it isn't a good solution to recommend).
+
+For instance, adding a small function to an existing opaque
+datastructure (Set, Map, Stack, Buffer) would generally induce a very
+small maintenance cost and it's not really possible to do that outside
+the stdlib (short of completely reimplementing the datastructure
+altogether). The argument is weaker for extending e.g. List or Array,
+since this can more easily be done outside stdlib. That said, if the
+addition is useful enough, there is no strong reason not to accept it.
+
+
+== More to discuss
+
+* When a module `M` has a main type it must be named `M.t`.
+* Functions should take the main type `t` as first argument (unless
+  there is a good reason for it, see next point).
+* If a functions takes another function as argument, that argument
+  should have the label `f` and not be last (to be able to omit the
+  label).
+* When a function may raise an exception as a non-exceptional
+  behavior, its name must have the suffix `_exn` to remind the user
+  that the exception should be handled.  Another function returning an
+  option instead must be provided.  Example:
+[source,ocaml]
+----
+val find : 'a t -> key -> 'a option
+val find_exn : 'a t -> key -> 'a
+----
+
+
+=== Collections
+
+* If a module `M` defines a collection of values, say `'a M.t`, it
+  must also define the following standard higher order functions:
+
+[source,ocaml]
+----
+val iter : f:('a -> unit) -> 'a t -> unit
+val iteri : f:(int -> 'a -> unit) -> 'a t -> unit
+val map : f:('a -> 'b) -> 'a t -> 'b t
+val mapi : f:(int -> 'a -> 'b) -> 'a t -> 'b t
+val fold : f:('a -> 'b -> 'a) -> init:'a -> 'b t -> 'a
+----
+
+The `iter`, `map` and `fold` function need not to specify the order in
+which arguments are passed to `f`.
+
+* Collections should be convertible to and from sequences by
+  providing:
+[source,ocaml]
+----
+val to_seq : 'a t -> 'a Seq.t
+val of_seq : 'a Seq.t -> 'a t
+----
+
+=== Arithmetic
+
+* When it makes sense, a module should provide _both_ standard infix
+  operators and named functions:
+
+[source,ocaml]
+----
+val neg : t -> t (* unary negation *)
+val add : t -> t -> t (* addition *)
+val sub : t -> t -> t (* subtraction *)
+val mul : t -> t -> t (* multiplication *)
+val div : t -> t -> t (* division *)
+val ( ~+ ) : t -> t (* identity *)
+val ( ~- ) : t -> t (* unary negation *)
+val ( + ) : t -> t -> t
+val ( - ) : t -> t -> t
+val ( * ) : t -> t -> t
+val ( / ) : t -> t -> t
+val ( ** ) : t -> int -> t (* exponentiation *)
+----
+
+The type of exponents of `( ** )` may be more general than `int`.
+
+
+=== Ordering
+
+* Unless it makes no sense, modules must provide a function
+[source,ocaml]
+----
+val compare : t -> t -> int
+----
+that defines a _total_ order on the main type `t`.  This is
+necessary to be able to build collections using `M.t` as the key.
```

Comment:
I think an `equal` compatible with the result of `compare` should be added. Before we have the right tools  defining `hash` is [tricky](https://github.com/ocaml/ocaml/pull/1940#issuecomment-409668832). If you really want to add it then we should add the seeded hash signature.

Reply:
I'm not against adding also the seeded hash signature, but I want to be able to just do `Hashtbl.Make(Int32)`, and currently I can't.  For modules exposing a parametric type, this is slightly less useful since one needs extra boilerplate anyway to pass the "internal" (seeded) hash.

Reply:
@alainfrisch Yes, `hash` and `equal` (compatible with `compare`) would be useful too.